groth16-framework/test_data/Groth16VerifierExtensions.sol

```diff
@@ -136,54 +157,78 @@ contract Query2 is Verifier {
         require(
-            max_block_number == query.maxBlockNumber,
+            maxBlockNumber == query.maxBlockNumber,
             "The parsed max block number must be equal to the expected one in query."
         );
 
-        address contract_address = convertToAddress(pis, PI_CONTRACT_ADDR_OFFSET);
+        address contractAddress = convertToAddress(pis, PI_CONTRACT_ADDR_OFFSET);
         require(
-            contract_address == query.contractAddress,
+            contractAddress == query.contractAddress,
             "The parsed contract address must be equal to the expected one in query."
         );
 
-        address user_address = convertToAddress(pis, PI_USER_ADDR_OFFSET);
+        address userAddress = convertToAddress(pis, PI_USER_ADDR_OFFSET);
         require(
-            user_address == query.userAddress,
+            userAddress == query.userAddress,
             "The parsed user address must be equal to the expected one in query."
         );
 
-        bytes32 block_hash = convertToHash(pis, PI_BLOCK_HASH_OFFSET);
+        bytes32 blockHash = bytes32(convertToHash(pis, PI_BLOCK_HASH_OFFSET));
         require(
-            block_hash == query.blockHash,
+            blockHash == query.blockHash,
             "The parsed block hash must be equal to the expected one in query."
         );
+
+        if (identifier == QUERY_IDENTIFIER_ERC20) {
+            uint256 rewardsRate = convertByteSliceToU256(pis, PI_REWARDS_RATE_OFFSET);
+            require(
+                rewardsRate == query.rewardsRate,
+                "The parsed rewards rate must be equal to the expected one in query."
+            );
+        }
+    }
+
+    // Parse the query identifier.
+    function parseQueryIdentifier(bytes memory pis) internal pure returns (uint8) {
+        return uint8(pis[PI_QUERY_IDENTIFIER_OFFSET]);
+    }
+
+    // Parse the query result from the plonky2 public inputs.
+    function parseQueryResult(bytes memory pis, uint8 identifier) internal pure returns (uint256[] memory) {
+        if (identifier == QUERY_IDENTIFIER_NFT) {
+            return parseNftIds(pis);
+        } else if (identifier == QUERY_IDENTIFIER_ERC20) {
+            return parseErc20Result(pis);
+        } else {
+            revert("Unsupported query identifier");
+        }
     }
 
     // Parse the `L` NFT IDs from the plonky2 public inputs.
     function parseNftIds(bytes memory pis) internal pure returns (uint256[] memory) {
         uint256[] memory nft_ids = new uint256[](L);
         for (uint32 i = 0; i < L; ++i) {
-            nft_ids[i] = uint256(convertToLeftPaddingU32(pis, PI_NFT_IDS_OFFSET + i * 8));
+            // TODO: need to test more in lagrange-lpn-contracts, it works with the groth16 test (decoding from the output by ABI).
+            nft_ids[i] = uint256(convertToU32(pis, PI_NFT_IDS_OFFSET + i * 8));
+            // nft_ids[i] = uint256(convertToLeftPaddingU32(pis, PI_NFT_IDS_OFFSET + i * 8));
```

Comment:
Leftover?

Reply:
Deleted in commit [7f58f59](https://github.com/Lagrange-Labs/Euclid-database/pull/21/commits/7f58f59a48096cd19872b80395c947591840f5ba).